File: php/tests/CarRelationshipsSampleTest.php
<?php

declare(strict_types=1);

namespace Oojs\Tests;

use Oojs\Registry;
use PHPUnit\Framework\TestCase;

use function Oojs\validate;

/**
 * Sample schema + instance: horizontal has-one / has-many via IDs.
 */
class CarRelationshipsSampleTest extends TestCase
{
    public function test_car_relationships_sample_valid(): void
    {
        $schema = [
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/car-rel',
            'types' => [
                'Person' => [
                    'properties' => [
                        'personId' => ['type' => 'string'],
                        'name'     => ['type' => 'string'],
                        // has-many: list of car IDs owned/managed by the person
                        'carIds'   => ['type' => 'array', 'items' => ['type' => 'string']],
                    ],
                    'required' => ['personId', 'name'],
                ],
                'Car' => [
                    'properties' => [
                        'carId'   => ['type' => 'string'],
                        'make'    => ['type' => 'string'],
                        'model'   => ['type' => 'string'],
                        // has-one: owner is referenced by ID (not embedded)
                        'ownerId' => ['type' => 'string'],
                        // has-one: current garage by ID (not embedded)
                        'garageId' => ['type' => 'string'],
                    ],
                    'required' => ['carId', 'make', 'model', 'ownerId'],
                ],
                'Garage' => [
                    'properties' => [
                        'garageId' => ['type' => 'string'],
                        'name'     => ['type' => 'string'],
                        // has-many: list of car IDs stored in the garage
                        'carIds'   => ['type' => 'array', 'items' => ['type' => 'string']],
                    ],
                    'required' => ['garageId', 'name'],
                ],
                'Fleet' => [
                    'properties' => [
                        'fleetId'   => ['type' => 'string'],
                        'name'      => ['type' => 'string'],
                        // has-many: list of car IDs in the fleet
                        'carIds'    => ['type' => 'array', 'items' => ['type' => 'string']],
                        // has-many: list of people IDs associated with the fleet
                        'personIds' => ['type' => 'array', 'items' => ['type' => 'string']],
                    ],
                    'required' => ['fleetId', 'name'],
                ],
            ],
        ];

        $instance = [
            '_type' => 'Fleet',
            'fleetId' => 'fleet-1',
            'name' => 'City Fleet',
            'carIds' => ['car-1', 'car-2'],
            'personIds' => ['person-1'],
        ];

        $r = new Registry();
        $loaded = $r->loadDict($schema);
        $errs = validate($instance, $loaded->types['Fleet'], $loaded, $r);

        $this->assertSame([], $errs);
    }
}

File: java/src/test/java/org/oojs/CarRelationshipsSampleTest.java
package org.oojs;

import org.junit.jupiter.api.Test;
import org.oojs.model.Schema;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;

class CarRelationshipsSampleTest extends TestHelpers {

    @Test
    void car_relationships_sample_valid() {
        Map<String, Object> schemaMap = map(
                "$oojs", "1.0",
                "$id", "https://example.org/schemas/car-rel",
                "types", map(
                        "Person", map(
                                "properties", map(
                                        "personId", map("type", "string"),
                                        "name", map("type", "string"),
                                        "carIds", map("type", "array", "items", map("type", "string"))),
                                "required", list("personId", "name")),
                        "Car", map(
                                "properties", map(
                                        "carId", map("type", "string"),
                                        "make", map("type", "string"),
                                        "model", map("type", "string"),
                                        "ownerId", map("type", "string"),
                                        "garageId", map("type", "string")),
                                "required", list("carId", "make", "model", "ownerId")),
                        "Garage", map(
                                "properties", map(
                                        "garageId", map("type", "string"),
                                        "name", map("type", "string"),
                                        "carIds", map("type", "array", "items", map("type", "string"))),
                                "required", list("garageId", "name")),
                        "Fleet", map(
                                "properties", map(
                                        "fleetId", map("type", "string"),
                                        "name", map("type", "string"),
                                        "carIds", map("type", "array", "items", map("type", "string")),
                                        "personIds", map("type", "array", "items", map("type", "string"))),
                                "required", list("fleetId", "name"))));

        Map<String, Object> instance = map(
                "_type", "Fleet",
                "fleetId", "fleet-1",
                "name", "City Fleet",
                "carIds", list("car-1", "car-2"),
                "personIds", list("person-1"));

        Registry r = new Registry();
        Schema schema = r.loadMap(schemaMap);
        List<ValidationError> errs = new Validator(r).validate(instance, schema.types.get("Fleet"), schema);

        assertEquals(List.of(), errs);
    }
}
